File: Java/LocalMinimum.java
import java.util.Arrays;

public class LocalMinimum {
	public static void main(String[] args) {
		int[][] arr = {{7,12,14,22},{8,4,6,6},{6,5,6,7},{5,2,10,6}};
		for( int i=0;i<4;i++ ) {
			for( int j=0;j<4;j++ ) {
				System.out.print(arr[i][j]+"\t");  
			}  
			System.out.println();  
		}  
		// System.out.println("Original Array:"+Arrays.deepToString(arr));

		Minimum local_minimum = find_local_minimum(arr,0,arr.length-1,0,arr[0].length-1);
		System.out.println("Local Minimum: ("+local_minimum.x+","+local_minimum.y+")");

	}

	public static Minimum find_local_minimum(int[][] arr,int ll,int hl,int lw,int hw) {
		// System.out.println(""+ll+hl+lw+hw);
		int n = (ll+hl)/2;
		int m = (lw+hw)/2;

		int[] middle_row = arr[n];
		int[] middle_column = new int[arr.length];
		for ( int i=0;i<arr.length;i++ ) {
			middle_column[i] = arr[i][m];
		}
		// System.out.println("Middle row:"+Arrays.toString(middle_row));
		// System.out.println("Middle column:"+Arrays.toString(middle_column));
		Minimum min = find_minimum(middle_row,middle_column,n,m);
		// System.out.println(""+min.x+min.y+arr[min.x][min.y]);
		int x = min.x;
		int y = min.y;

		// 8 boundary conditions
		if ( x == 0 && y == 0 ) {
			if ( arr[x][y] <= arr[x+1][y] && arr[x][y] <= arr[x][y+1]) return min;
			else return new Minimum(-1,-1);
		} else if ( x == hl && y == hw ) {
			if ( arr[x][y] <= arr[x-1][y] && arr[x][y] <= arr[x][y-1] )	return min;
			else return new Minimum(-1,-1);
		} else if ( x == hl && y == 0 ) {
			if ( arr[x][y] <= arr[x-1][y] && arr[x][y] <= arr[x][y+1] ) return min;
			else return new Minimum(-1,-1) ;
		} else if ( x == 0 && y == hw ) {
			if ( arr[x][y] <= arr[x+1][y] && arr[x][y] <= arr[x][y-1] )	return min;
			else return new Minimum(-1,-1);
		} else if ( x == 0 ) {
			if ( arr[x][y] <= arr[x+1][y] && arr[x][y] <= arr[x][y+1] && arr[x][y] <= arr[x][y-1] ) return min;
			else return new Minimum(-1,-1);
		} else if ( y == 0 ) {
			if ( arr[x][y] <= arr[x-1][y] && arr[x][y] <= arr[x+1][y] && arr[x][y] <= arr[x][y+1] )	return min;
			else return new Minimum(-1,-1);
		} else if ( x == hl ) {
			if ( arr[x][y] <= arr[x-1][y] && arr[x][y] <= arr[x][y+1] && arr[x][y] <= arr[x][y-1] )	return min;
			else return new Minimum(-1,-1);
		} else if ( y == hw ) {
			if ( arr[x][y] <= arr[x-1][y] && arr[x][y] <= arr[x+1][y] && arr[x][y] <= arr[x][y-1] ) return min;
			else return new Minimum(-1,-1);
		}
		
		if ( x>0 && arr[x][y] <= arr[x-1][y] ) {
			if ( x<hl && arr[x][y] <= arr[x+1][y] ) {
				if ( y>0 && arr[x][y] <= arr[x][y-1] ) {
					if ( y<hw && arr[x][y] <= arr[x][y+1] ) {
						return min;
					} else {
						if ( x>n ) return find_local_minimum(arr,n,hl,m,hw);
						else return find_local_minimum(arr,ll,n,m,hw);
					} 
				} else {
					if ( x>n ) return find_local_minimum(arr,n,hl,lw,m);
					else return find_local_minimum(arr,ll,n,lw,m);
				}
			} else  {
				if ( y>m ) return find_local_minimum(arr,n,hl,m,hw);
				else return find_local_minimum(arr,n,hl,lw,m);
			}
		} else {
			if ( y>m ) return find_local_minimum(arr,ll,n,m,hw);
			else return find_local_minimum(arr,ll,n,lw,m);
		}
	}

	public static Minimum find_minimum(int[] arr1,int[] arr2,int n,int m) {
		int l1 = arr1.length;
		int l2 = arr2.length;
		int l = l1+l2;
		int[] arr = new int[l];
		System.arraycopy(arr1, 0, arr, 0, l1);
   		System.arraycopy(arr2, 0, arr, l1, l2);
   		Minimum minimum_pos = new Minimum(n,arr[0]);
   		int minimum = arr[0];
   		// System.out.println("Combine Array:"+Arrays.toString(arr));

   		for ( int i=1;i<l;i++ ) {
   			if ( minimum > arr[i] ) {
   				minimum = arr[i];
   				if ( i<l1 ) minimum_pos = new Minimum(n,i);
   				else minimum_pos = new Minimum(i-l1,m);
   			}
   		}
   		
   		return minimum_pos;
	}
}

class Minimum {
	public final int x;
	public final int y;

	Minimum(int x,int y) {
		this.x = x;
		this.y = y;
	}
}

File: Python/local_minimum.py
import numpy as np
def find_local_minimum(arr,ll,hl,lw,hw):
	
	n = (ll+hl)/2
	m = (lw+hw)/2

	(x,y) = find_min(arr[n,0:hw+1],arr[0:hl+1,m],n,m)
	# print arr[x,y] 
	
	# print ll,hl,lw,hw
	
	# print x,y

	#8 boundary conditions
	if x == 0 and y == 0:
		if arr[x][y] <= arr[x+1][y] and arr[x][y] <= arr[x][y+1]:
			return (x,y)
		else:
			return -1

	elif x == hl and y == hw:
		if arr[x][y] <= arr[x-1][y] and arr[x][y] <= arr[x][y-1]:
			return (x,y)
		else:
			return -1

	elif x == hl and y == 0:
		if arr[x][y] <= arr[x-1][y] and arr[x][y] <= arr[x][y+1]:
			return (x,y)
		else:
			return -1 

	elif x == 0 and y == hw:
		if arr[x][y] <= arr[x+1][y] and arr[x][y] <= arr[x][y-1]:
			return (x,y)
		else:
			return -1

	elif x == 0:
		if arr[x][y] <= arr[x+1][y] and arr[x][y] <= arr[x][y+1] and arr[x][y] <= arr[x][y-1]:
			return (x,y)
		else:
			return -1

	elif y == 0:
		if arr[x][y] <= arr[x-1][y] and arr[x][y] <= arr[x+1][y] and arr[x][y] <= arr[x][y+1]:
			return (x,y)
		else:
			return -1

	elif x == hl:
		if arr[x][y] <= arr[x-1][y] and arr[x][y] <= arr[x][y+1] and arr[x][y] <= arr[x][y-1]:
			return (x,y)
		else:
			return -1

	elif y == hw:
		if arr[x][y] <= arr[x-1][y] and arr[x][y] <= arr[x+1][y] and arr[x][y] <= arr[x][y-1]:
			return (x,y)
		else:
			return -1
	
	if x>0 and arr[x][y] <= arr[x-1][y]:
		if x<hl and arr[x][y] <= arr[x+1][y]:
			if y>0 and arr[x][y] <= arr[x][y-1] :
				if y<hw and arr[x][y] <= arr[x][y+1]:
					return (x,y)
				else:
					if x>n:
						return find_local_minimum(arr,n,hl,m,hw)
					else:
						return find_local_minimum(arr,ll,n,m,hw)
			else:
				if x>n:
						return find_local_minimum(arr,n,hl,lw,m)
				else:
					return find_local_minimum(arr,ll,n,lw,m)
		else:
			if y>m:
				return find_local_minimum(arr,n,hl,m,hw)
			else:
				return find_local_minimum(arr,n,hl,lw,m)
	else:
		if y>m:
			return find_local_minimum(arr,ll,n,m,hw)
		else:
			return find_local_minimum(arr,ll,n,lw,m)

def find_min(arr1,arr2,n,m) :
	minimum = arr1[0]
	l1=len(arr1)
	l2=len(arr2)
	arr = np.concatenate([arr1,arr2])
	l=len(arr)

	# arr1.extend(arr2)
	for i in range(l):
		if minimum > arr[i]:
			minimum = arr[i]
			if i < l1:
				minimum_pos = (n,i)
			else:
				minimum_pos = (i-l1,m)
	return minimum_pos

if __name__ == '__main__':
	arr = [[7,12,14,22],[8,4,2,6],[6,5,6,7],[5,6,10,6]]
	print 'Original: \n',np.matrix(arr)
	arr = np.array(arr)
	# print np.shape(arr)[0]
	# print np.shape(arr)[1]
	local_minimum = find_local_minimum(arr,0,np.shape(arr)[0]-1,0,np.shape(arr)[1]-1)
	print 'Local Minimum: ',local_minimum